cpp/src/arrow/array.h

```diff
@@ -734,6 +745,27 @@ class ARROW_EXPORT DictionaryArray : public Array {
  private:
   void SetData(const std::shared_ptr<ArrayData>& data);
 
+  /// \brief Check if all indices are within valid range
+  ///
+  /// \param[in] indices dictionary indices
+  /// \param[in] range valid range of indices (0 <= index < range)
+  template <typename ArrowType>
+    static bool SanityCheck(const std::shared_ptr<Array>& indices, const int64_t range) {
+    using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+    std::shared_ptr<ArrayType> array = std::static_pointer_cast<ArrayType>(indices);
+    const typename ArrowType::c_type* data = array->raw_values();
+    const int64_t size = sizeof(data) / sizeof(data[0]);
+
+    for (int64_t idx = 0; idx < size; ++idx) {
+      if (!array->IsNull(idx)) {
+        if (data[idx] < 0 || data[idx] >= range) {
+	  return false;
+        }
+      }
+    }
+    return true;
+  }
```

Comment:
Not sure if we need a new member function; can put this in the compilation unit

Reply:
Do you mean moving SanityCheck() out of DictionaryArray class and making it a standalone template function under arrow namespace?

Reply:
yes, or in an anonymous namespace would be fine